src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,184 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
-                {
-                    elementType = this.GetSpecialType(SpecialType.System_Char, diagnostics, initializerSyntax);
-                    Debug.Assert(!elementType.IsManagedType);
-                }
-                else if (initializerType.IsArray())
-                {
-                    // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
-                    var arrayType = (ArrayTypeSymbol)initializerType;
-                    elementType = arrayType.ElementType;
+                case BoundKind.FieldAccess:
+                    var fa = (BoundFieldAccess)initializerOpt;
+                    if (fa.FieldSymbol.IsFixed)
+                    {
+                        elementType = ((PointerTypeSymbol)fa.Type).PointedAtType;
+                        break;
+                    }
+
+                    goto default;
+
+                default:
+                    //  fixed (T* variable = <expr>) ...
 
-                    if (elementType.IsManagedType)
+                    // check for arrays
+                    if (initializerType.IsArray())
                     {
-                        Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
-                        hasErrors = true;
+                        // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
+                        var arrayType = (ArrayTypeSymbol)initializerType;
+                        elementType = arrayType.ElementType;
+                        break;
                     }
-                }
-                else if (!initializerOpt.HasAnyErrors)
-                {
-                    elementType = initializerOpt.Type;
-                    switch (initializerOpt.Kind)
+
+                    // check for a special ref-returning method
+                    var additionalDiagnostics = DiagnosticBag.GetInstance();
+                    try
                     {
-                        case BoundKind.AddressOfOperator:
-                            elementType = ((BoundAddressOfOperator)initializerOpt).Operand.Type;
+                        getPinnableMethod = GetDangerousGetPinnableReferenceMethod(initializerOpt, additionalDiagnostics);
+
+                        bool extensisbleFixedEnabled = ((CSharpParseOptions)initializerOpt.SyntaxTree.Options)?.IsFeatureEnabled(MessageID.IDS_FeatureExtensibleFixedStatement) != false;
+
+                        // check for String
+                        // NOTE: We will allow DangerousGetPinnableReferenceMethod to take precendence, but only if it is a member of System.String
+                        if (initializerType.SpecialType == SpecialType.System_String &&
+                            ((object)getPinnableMethod == null || !extensisbleFixedEnabled || getPinnableMethod.ContainingType.SpecialType != SpecialType.System_String))
```

Comment:
>extensisbleFixedEnabled [](start = 67, length = 23)

I don't think the presence/absence of the feature flag should affect binding. Only diagnostics should be conditional on the feature flag.

Reply:
This is a very special case. We want that extensible pinnable to win over when implemented on System.String. 

---
In reply to: [167648804](https://github.com/dotnet/roslyn/pull/24469#discussion_r167648804) [](ancestors = 167648804)

Reply:
You a right though - we should not check extensisbleFixedEnabled  here. 
The implementors of System.String will guarantee backwards compatibility of the new pattern. So this is more a matter of runtime versioning, not compiler versioning.

---
In reply to: [169196056](https://github.com/dotnet/roslyn/pull/24469#discussion_r169196056) [](ancestors = 169196056,167648804)